app/commands/solution/search_community_solutions.rb

```diff
@@ -17,25 +17,64 @@ def initialize(exercise, page: nil, per: nil, criteria: nil)
     end
 
     def call
-      @solutions = exercise.solutions.published
+      results = client.search(index: 'solutions', body: search_body)
 
-      filter_criteria!
-      sort!
+      solution_ids = results["hits"]["hits"].map { |hit| hit["_source"]["id"] }
+      solutions = Solution.where(id: solution_ids).
+        includes(:exercise, :track).
+        order(Arel.sql("FIND_IN_SET(id, '#{solution_ids.join(',')}')")).
+        to_a
 
-      @solutions.page(page).per(per)
+      total_count = results["hits"]["total"]["value"].to_i
+      Kaminari.paginate_array(solutions, total_count: total_count).
+        page(page).per(per)
     end
 
-    def filter_criteria!
-      return if @criteria.blank?
+    private
+    attr_reader :exercise, :per, :page, :solutions, :criteria
+
+    def search_body
+      {
+        query: search_query,
+        sort: search_sort,
 
-      @solutions = @solutions.joins(:user).where("users.handle LIKE ?", "%#{criteria}%")
+        # Only return the solution IDs, not the entire document, to improve performance
+        _source: [:id],
+
+        # Paging information
+        from: (page - 1) * per,
+        size: per
```

Comment:
Just checked above to make sure these weren't ever possibly nil and noticed you could remove the `....present? && ` checks from L14 and L15 too if you fancied it.